docs/design/variadics.md

```diff
@@ -0,0 +1,1018 @@
+the scrutinee segments to obtain `(... ⟬Vector(each T), Vector(i32)⟭)`. This has
+a single segment with arity `‖each T‖+1`, which can match `‖each Next‖+1`
+because the deduced arity `‖each Next‖` behaves as a deduced parameter of the
+pattern, so they match by deducing `‖each Next‖ == ‖each T‖`.
+
+When merging segments of the scrutinee, we don't attempt to form name packs and
+replace them with invented names, but we also don't need to: we don't require a
+merged scrutinee segments to have a single scalar component.
+
+The search for this rewrite processes each pattern segment to the left of the
+segment with deduced arity, in order from left to right. For each pattern
+segment, it greedily merges unmatched scrutinee segments from left to right
+until their cumulative shape is greater than or equal to the shape of the
+pattern segment, and then splits off a scrutinee segment on the right if
+necessary to make the shapes exactly match. Pattern segments to the right of the
+segment with deduced arity are processed the same way, but with left and right
+reversed, so that segments are always processed from the outside in.
+
+See the [appendix](#appendix-type-system-formalism) for the rewrite rules that
+govern merging and splitting.
+
+Once we have the pattern and scrutinee segments in one-to-one correspondence, we
+check each scalar component of the scrutinee type against the scalar component
+of the corresponding pattern type segment (by construction, the pattern type
+segment has only one scalar component). Since we are checking scalar components
+against scalar components, this proceeds according to the usual rules of
+non-variadic typechecking.
+
+> **TODO:** Extend this approach to fall back to a complementary approach, where
+> the pattern and scrutinee trade roles: we maximally merge the scrutinee tuple,
+> while requiring each segment to have a single scalar component, and then
+> merge/split the pattern tuple to match it, without requiring pattern tuple
+> segments to have a single scalar component. This isn't quite symmetric with
+> the current approach, because when processing the scrutinee we can't merge
+> deduced parameters (scrutinees don't have any), but we can invent new `let`
+> bindings.
+
+## Appendix: Type system formalism
+
+A _pack literal_ is a comma-separated sequence of segments, enclosed in `⟬⟭`
+delimiters. A pack literal can appear in an expression, pattern, or name
+context, and every segment must be valid in the context where the pack literal
+appears (for example, the segments of a pack literal in a name context must all
+be names). Pack literals cannot be nested, and cannot appear outside a pack
+expansion.
+
+### Explicit deduced arities
+
+In this formalism, deduced arities are explicit rather than implicit, so Carbon
+code must be desugared into this formalism as follows:
+
+For each pack expansion pattern, we introduce a binding pattern `__N:! Arity` as
```

Comment:
Is this always done, or only if needed? It seems like it might not be needed, for example when it has fixed arity (and so there is an earlier arity parameter that governs the arity of this pack expansion pattern).

(see your example at the end of this doc)